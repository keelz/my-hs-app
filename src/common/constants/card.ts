const BASE_URI = 'https://art.hearthstonejson.com/v1/render/latest';

const RENDER_PATH = '/{LOCALE}/{RESOLUTION}x/{ID}.{EXT}';

export default {
  BASE_URI,
  RENDER_PATH,
};
