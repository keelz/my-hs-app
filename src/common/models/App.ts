
export interface IComponentProps {
  className?: string | string[];
}
