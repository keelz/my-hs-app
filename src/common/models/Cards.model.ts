
// TODO: finish modeling card interface.
export interface ICard {
  artist?: string;
  cardClass: CardClassName;
  collectible: boolean;
  cost: number;
  dbfId: number;
  flavor: string;
  id: string;
  name: string;
  race?: Race;
  rarity?: Rarity;
  set: CardSet;
  text: string;
  type: CardType;
}

export enum CardClassName {
  DEATHKNIGHT = 'DEATHKNIGHT',
  DREAM = 'DREAM',
  DRUID = 'DRUID',
  HUNTER = 'HUNTER',
  MAGE = 'MAGE',
  NEUTRAL = 'NEUTRAL',
  PALADIN = 'PALADIN',
  PRIEST = 'PRIEST',
  ROGUE = 'ROGUE',
  SHAMAN = 'SHAMAN',
  WARLOCK = 'WARLOCK',
  WARRIOR = 'WARRIOR',
  WHIZBANG = 'WHIZBANG',
}

export enum CardExt {
  PNG = 'png',
}

export enum CardLocale {
  EN = 'enUS',
}

export enum CardResolution {
  SMALL = 256,
  LARGE = 512,
}

export enum CardSet {
  INVALID = 'INVALID',
	TEST_TEMPORARY = 'TEST_TEMPORARY',
	CORE = 'CORE',
	EXPERT1 = 'EXPERT1',
	HOF = 'HOF',
	MISSIONS = 'MISSIONS',
	DEMO = 'DEMO',
	NONE = 'NONE',
	CHEAT = 'CHEAT',
	BLANK = 'BLANK',
	DEBUG_SP = 'DEBUG_SP',
	PROMO = 'PROMO',
	NAXX = 'NAXX',
	GVG = 'GVG',
	BRM = 'BRM',
	TGT = 'TGT',
	CREDITS = 'CREDITS',
	HERO_SKINS = 'HERO_SKINS',
	TB = 'TB',
	SLUSH = 'SLUSH',
	LOE = 'LOE',
	OG = 'OG',
	OG_RESERVE = 'OG_RESERVE',
	KARA = 'KARA',
	KARA_RESERVE = 'KARA_RESERVE',
	GANGS = 'GANGS',
	GANGS_RESERVE = 'GANGS_RESERVE',
	UNGORO = 'UNGORO',
	ICECROWN = 'ICECROWN',
	LOOTAPALOOZA = 'LOOTAPALOOZA',
	GILNEAS = 'GILNEAS',
	BOOMSDAY = 'BOOMSDAY',
	TROLL = 'TROLL',
	TAVERNS_OF_TIME = 'TAVERNS_OF_TIME',
}

export enum CardType {
  INVALID = 'INVALID',
	GAME = 'GAME',
	PLAYER = 'PLAYER',
	HERO = 'HERO',
	MINION = 'MINION',
	SPELL = 'SPELL',
	ENCHANTMENT = 'ENCHANTMENT',
	WEAPON = 'WEAPON',
	ITEM = 'ITEM',
	TOKEN = 'TOKEN',
	HERO_POWER = 'HERO_POWER',
}

export enum Race {
  INVALID = 'INVALID',
	BLOODELF = 'BLOODELF',
	DRAENEI = 'DRAENEI',
	DWARF = 'DWARF',
	GNOME = 'GNOME',
	GOBLIN = 'GOBLIN',
	HUMAN = 'HUMAN',
	NIGHTELF = 'NIGHTELF',
	ORC = 'ORC',
	TAUREN = 'TAUREN',
	TROLL = 'TROLL',
	UNDEAD = 'UNDEAD',
	WORGEN = 'WORGEN',
	GOBLIN2 = 'GOBLIN2',
	MURLOC = 'MURLOC',
	DEMON = 'DEAMON',
	SCOURGE = 'SCOURGE',
	MECHANICAL = 'MECHANICAL',
	ELEMENTAL = 'ELEMENTAL',
	OGRE = 'OGRE',
	BEAST = 'BEAST',
	TOTEM = 'TOTEM',
	NERUBIAN = 'NERUBIAN',
	PIRATE = 'PIRATE',
	DRAGON = 'DRAGON',
	BLANK = 'BLANK',
	ALL = 'ALL',
	EGG = 'EGG',
  PET = 'PET',
}

export enum Rarity {
  INVALID = 'INVALID',
	COMMON = 'COMMON',
	FREE = 'FREE',
	RARE = 'RARE',
	EPIC = 'EPIC',
	LEGENDARY = 'LEGENDARY',
  UNKNOWN_6 = 'UNKNOWN_6',
}
