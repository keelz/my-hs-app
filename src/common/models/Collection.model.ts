import { ICard } from './Cards.model';

export interface ICollection {
  cards: ICard[];
}
