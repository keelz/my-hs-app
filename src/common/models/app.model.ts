import { MiddlewareAPI, Dispatch } from 'redux';
import { IActionType, IRootState } from '../../redux/Types';

export interface IComponentProps {
  className?: string | string[];
}

export interface IPagination {
  currentPage: number;
  itemsPerPage: number;
  pages: number;
  total: number;
}

export type AppMiddlewareApi = MiddlewareAPI<Dispatch<IActionType<any>>, IRootState>;

// todo: factor this to redux namespace.
export type MiddlewareOperation = (api: AppMiddlewareApi) =>
  (next: Dispatch<IActionType<any>>) =>
  (action: IActionType<any>) => void;

export enum BlockOrientation {
  TOP,
  RIGHT,
  BOTTOM,
  LEFT,
}
