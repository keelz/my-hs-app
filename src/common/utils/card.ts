import CARD from '../constants/card';
import HSJSON from '../constants/hsJson';
import {
  Card,
  CardClassName
} from '../../redux/reducers/Cards';

export const cardSrcWithParameters = (
  id: string,
  locale: string,
  resolution: number,
  ext: string
) => CARD.BASE_URI + CARD.RENDER_PATH
  .replace(`\{${HSJSON.REQUEST_PARAMS.LOCALE}\}`, locale)
  .replace(`\{${HSJSON.REQUEST_PARAMS.RESOLUTION}\}`, resolution.toString())
  .replace(`\{${HSJSON.REQUEST_PARAMS.ID}\}`, id)
  .replace(`\{${HSJSON.REQUEST_PARAMS.EXT}\}`, ext);

export const composeCardClassNames = (cards: Card[]): string[] => cards.reduce(
  (a: string[], b: Card) => {
    const { cardClass } = b;
    switch (cardClass) {
      case CardClassName.DEATHKNIGHT:
      case CardClassName.DREAM:
      case CardClassName.NEUTRAL:
      case CardClassName.WHIZBANG:
      case undefined: return a;
    }
    if (a.indexOf(cardClass) >= 0) return a;
    return [...a, cardClass].sort();
  },
  []
) as string[];

export default {
  cardSrcWithParameters,
  composeCardClassNames,
};
