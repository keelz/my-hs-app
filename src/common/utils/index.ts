import * as classnames from 'classnames';

export const composeClassname = (defaultClassname: string | string[]) =>
  (className?: string | string[]) =>
  classnames(defaultClassname, className);

export default {
  composeClassname,
};
