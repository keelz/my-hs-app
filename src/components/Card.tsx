import * as React from 'react';
import { composeAssetSource } from '../common/services/hsJsonApi';
import {
  CardExt,
  CardLocale,
  CardResolution
} from '../common/models/Card';

interface ICardProps {
  id: string;
  ext: CardExt;
  locale: CardLocale;
  resolution: CardResolution;
}

const Card: React.SFC<ICardProps> = props =>
  <img
    width={props.resolution}
    style={{ marginBottom: -45 }}
    src={composeAssetSource(
      props.id,
      props.locale,
      props.resolution,
      props.ext
    )} />;

export default Card;
