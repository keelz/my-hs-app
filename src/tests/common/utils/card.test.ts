import { composeCardClassNames } from '../../../common/utils/card';
import mockCollection from '../../../common/mocks/collection';
import { ICard } from '../../../common/models/Card';

describe('card utilities', () => {
  it('composes card class names correctly', () => {
    const result = composeCardClassNames(mockCollection as ICard[]);
    expect(result).toMatchSnapshot();
  });
});
