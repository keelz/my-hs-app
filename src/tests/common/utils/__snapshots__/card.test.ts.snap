// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`card utilities composes card class names correctly 1`] = `
Array [
  "DRUID",
  "HUNTER",
  "MAGE",
  "PALADIN",
  "PRIEST",
  "ROGUE",
  "SHAMAN",
  "WARLOCK",
  "WARRIOR",
  "NEUTRAL",
]
`;
